Source/WebCore/editing/MarkupAccumulator.cpp

```diff
@@ -217,18 +212,39 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
         namespaceHash.set(XMLNames::xmlNamespaceURI->impl(), xmlAtom().impl());
     }
 
-    if (root == SerializedNodes::SubtreeIncludingNode)
-        startAppendingNode(targetNode, &namespaceHash);
-
-    if (targetNode.document().isHTMLDocument() && elementCannotHaveEndTag(targetNode))
-        return;
-
-    Node* current = targetNode.hasTagName(templateTag) ? downcast<HTMLTemplateElement>(targetNode).content().firstChild() : targetNode.firstChild();
-    for ( ; current; current = current->nextSibling())
-        serializeNodesWithNamespaces(*current, SerializedNodes::SubtreeIncludingNode, &namespaceHash, tagNamesToSkip);
+    const Node* current = &targetNode;
+    do {
+        bool shouldTraverseChildren = true;
+        if (tagNamesToSkip && is<Element>(current)) {
+            for (auto& name : *tagNamesToSkip) {
+                if (downcast<Element>(current)->hasTagName(name))
+                    shouldTraverseChildren = false;
+            }
+        }
 
-    if (root == SerializedNodes::SubtreeIncludingNode)
-        endAppendingNode(targetNode);
+        bool shouldAppendNode = !(current == &targetNode && root != SerializedNodes::SubtreeIncludingNode);
+        if (shouldAppendNode)
+            startAppendingNode(*current, &namespaceHash);
+
+        Node* firstChild = current->hasTagName(templateTag) ? downcast<HTMLTemplateElement>(current)->content().firstChild() : current->firstChild();
+        if (shouldTraverseChildren && firstChild)
+            current = firstChild;
+        else if (current != &targetNode) {
+            bool shouldCloseTag = !(targetNode.document().isHTMLDocument() && elementCannotHaveEndTag(*current));
+            if (shouldAppendNode && shouldCloseTag)
+                endAppendingNode(*current);
+            while (current != &targetNode) {
+                if (current->nextSibling()) {
+                    current = current->nextSibling();
+                    break;
+                }
+                current = current->parentNode();
+                if (is<TemplateContentDocumentFragment>(current))
```

Comment:
I might be missing something, but it looks like this check never appeared in the recursive version? Is there a expected/testable behavior change?

Reply:
We need to have this check because in the iterative version, we need to get out of the template element's content whereas in the recursive version, it happened implicitly by way of returning from a function call.